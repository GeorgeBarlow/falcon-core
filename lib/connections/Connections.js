'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keytar = require('keytar');

var _keytar2 = _interopRequireDefault(_keytar);

var _electronStore = require('electron-store');

var _electronStore2 = _interopRequireDefault(_electronStore);

var _util = require('util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }
// Manage saved connections to databases. Encrypts passwords


// We can't import electron in jest so electron-store won't work.
// We need to use 'conf' as a drop-in replacement for electron-store
// in the testing environment
const FinalStore = process.env.NODE_ENV === 'test' ? require('conf') // eslint-disable-line
: _electronStore2.default;

/**
 * This class is a general manager for falcon database connections.
 * It can be extended to fit the needs of specific databases. For
 * example, if a specific database requires encryption, the .get()
 * method can be modified
 */
class Connections {
  constructor() {
    this.store = new FinalStore({
      defaults: {
        connections: []
      }
    });
  }

  /**
   * @TODO
   */
  validateBeforeCreation(connection) {
    return _asyncToGenerator(function* () {
      const isFilePath = yield Promise.resolve().then(() => require('is-valid-path'));
      const Joi = yield Promise.resolve().then(() => require('joi'));
      const fs = yield Promise.resolve().then(() => require('fs'));

      const customJoi = Joi.extend(function (joi) {
        return {
          base: joi.string(),
          name: 'string',
          language: {
            file: 'needs to be a file',
            file_exists: 'does not exist'
          },
          rules: [{
            name: 'file',
            validate(params, value, state, options) {
              return !isFilePath(value) ? this.createError('string.file', { v: value, q: params.q }, state, options) : value;
            }
          }, {
            name: 'file_exists',
            validate(params, value, state, options) {
              console.log(value);
              return fs.existsSync(value) ? this.createError('string.file_exists', { v: value, q: params.q }, state, options) : value;
            }
          }]
        };
      });

      const schema = function () {
        switch (connection.type) {
          case 'sqlite':
            {
              return customJoi.object().keys({
                id: customJoi.string().required(),
                name: customJoi.string().required(),
                database: customJoi.string().file().file_exists().required(),
                type: customJoi.string()
              });
            }
          default:
            {
              throw new Error(`Unknown database type "${connection.type}". This probably means it is not supported`);
            }
        }
      }();

      const errors = customJoi.validate(connection, schema, { abortEarly: false });
      return errors.error.details;
    })();
  }

  create(connection) {
    var _this = this;

    return _asyncToGenerator(function* () {
      const connections = yield _this.getAll();
      connections.push(connection);
      _this.store.set('connections', connections);
    })();
  }

  /**
   * Delete a connection by it's id
   */
  delete(connectionId) {
    var _this2 = this;

    return _asyncToGenerator(function* () {
      const connections = yield _this2.getAll();
      const filteredConnections = connections.filter(function (connection) {
        return connection.id !== connectionId;
      });
      _this2.store.set('connections', filteredConnections);
    })();
  }

  deleteAll() {
    var _this3 = this;

    return _asyncToGenerator(function* () {
      yield _this3.store.set('connections', []);
    })();
  }

  /**
   * Update a connection by giving a new config
   */
  update(connectionId, connection) {
    var _this4 = this;

    return _asyncToGenerator(function* () {
      const connections = yield _this4.getAll();
      const connectionToUpdateIndex = connections.findIndex(function (conn) {
        return conn.id === connectionId;
      });

      switch (connectionToUpdateIndex) {
        case -1:
          {
            throw new Error(`Connection with id "${connectionId}" not found`);
          }
        default:
          {
            connections[connectionToUpdateIndex] = connection;
          }
      }

      _this4.store.set('connections', connections);
    })();
  }

  getAll() {
    var _this5 = this;

    return _asyncToGenerator(function* () {
      return _this5.store.get('connections');
    })();
  }

  get(connectionId) {
    var _this6 = this;

    return _asyncToGenerator(function* () {
      const connections = yield _this6.getAll();
      const connectionIndex = connections.findIndex(function (conn) {
        return conn.id === connectionId;
      });

      switch (connectionIndex) {
        case -1:
          {
            throw new Error(`Connection with id "${connectionId}" not found`);
          }
        default:
          {
            return connections[connectionIndex];
          }
      }
    })();
  }
}
exports.default = Connections;
//# sourceMappingURL=Connections.js.map