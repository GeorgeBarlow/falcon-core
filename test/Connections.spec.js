// @flow
/* eslint no-await-in-loop: 0 */
import path from 'path';
import Connections from '../src/config/ConnectionManager';

async function connectionFactory(connections, connectionCount: number = 1) {
  const array = new Array(connectionCount);
  for (let i = 0; i < array.length; i++) {
    array[i] = await connections.add({
      id: `test-id-${i + 1}`,
      type: 'sqlite',
      name: `test-connection-${i + 1}`,
      database: path.join(__dirname, 'databases/sqlite/demo.sqlite')
    });
  }
  return Promise.all(array);
}

function removeDatabaseProperty(object: Object): Object {
  const foo = { ...object };
  delete foo.database;
  assertDatabaseProperty(object);
  return foo;
}

function assertDatabaseProperty(object: Object): Object {
  expect(object).toHaveProperty('database');
  expect(object.database).toBeTruthy();
}

describe('Connections', function testConnections() {
  beforeEach(async () => {
    this.connections = new Connections();
    await connectionFactory(this.connections, 3);
  });

  afterEach(async () => {
    await this.connections.removeAll();
  });

  it('should get all connections', async () => {
    const connections = await this.connections.getAll();
    expect(connections.map(removeDatabaseProperty)).toMatchSnapshot();
  });

  it('should delete a single connection', async () => {
    const connections = await this.connections.getAll();
    const connectionIdToDelete = connections[0].id;
    await this.connections.remove(connectionIdToDelete);
    const newConnections = await this.connections.getAll();
    expect(newConnections.map(removeDatabaseProperty)).toMatchSnapshot();
  });

  it('should update a single connection', async () => {
    const connections = await this.connections.getAll();
    const connectionIdToDelete = connections[0].id;
    await this.connections.update(connectionIdToDelete, {
      id: 'test-id-foo',
      name: 'test-connection-foo',
      password: 'test-password',
      type: 'sqlite'
    });
    const newConnections = await this.connections.getAll();
    expect(newConnections.map(removeDatabaseProperty)).toMatchSnapshot();
  });

  it('should get a single connection', async () => {
    const connections = await this.connections.getAll();
    const connectionId = connections[2].id;
    const connection = await this.connections.get(connectionId);
    expect(removeDatabaseProperty(connection)).toMatchSnapshot();
  });

  it('should perform basic validation', async () => {
    try {
      await this.connections.add({
        id: 12,
        database: 'aJ@#LJ#@KL$KL@#sdf',
        type: 'sqlite'
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
    try {
      await this.connections.add({
        id: 12,
        database: '/usr/foo',
        type: 'sqlite'
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
    try {
      await this.connections.add({
        id: 12,
        database: '/usr/foo',
        type: 'sqlite'
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
    try {
      await this.connections.add({
        id: 'foo',
        name: 'foo',
        database: '/usr/foo',
        type: 'sqlite'
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
    try {
      await this.connections.add({
        id: 'foo',
        name: 'foo',
        database: '/usr/local/bin/npm',
        type: 'sqlite'
      });
    } catch (e) {
      expect(e).toMatchSnapshot();
    }
  });

  it('should check if a sqlite file is valid or not', async () => {
    const database = path.join(__dirname, 'databases/sqlite/demo.sqlite');
    const conn = await this.connections.add({
      id: 'foo',
      name: 'foo',
      type: 'sqlite',
      database
    });
    conn.data.item = removeDatabaseProperty(conn.data.item);
    expect(conn).toMatchSnapshot();
  });
});
