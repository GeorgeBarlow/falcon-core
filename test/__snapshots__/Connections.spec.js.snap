// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connections should delete a single connection 1`] = `
Array [
  Object {
    "id": "test-id-2",
    "name": "test-connection-2",
    "password": "test-password",
    "type": "sqlite",
  },
  Object {
    "id": "test-id-3",
    "name": "test-connection-3",
    "password": "test-password",
    "type": "sqlite",
  },
]
`;

exports[`Connections should get a single connection 1`] = `
Object {
  "id": "test-id-3",
  "name": "test-connection-3",
  "password": "test-password",
  "type": "sqlite",
}
`;

exports[`Connections should get all connections 1`] = `
Array [
  Object {
    "id": "test-id-1",
    "name": "test-connection-1",
    "password": "test-password",
    "type": "sqlite",
  },
  Object {
    "id": "test-id-2",
    "name": "test-connection-2",
    "password": "test-password",
    "type": "sqlite",
  },
  Object {
    "id": "test-id-3",
    "name": "test-connection-3",
    "password": "test-password",
    "type": "sqlite",
  },
]
`;

exports[`Connections should perform basic validation 1`] = `
Array [
  Object {
    "context": Object {
      "key": "id",
      "label": "id",
      "value": 12,
    },
    "message": "\\"id\\" must be a string",
    "path": Array [
      "id",
    ],
    "type": "string.base",
  },
  Object {
    "context": Object {
      "key": "name",
      "label": "name",
    },
    "message": "\\"name\\" is required",
    "path": Array [
      "name",
    ],
    "type": "any.required",
  },
  Object {
    "context": Object {
      "key": "database",
      "label": "database",
      "q": undefined,
      "v": "aJ@#LJ#@KL$KL@#sdf",
    },
    "message": "\\"database\\" needs to be a file",
    "path": Array [
      "database",
    ],
    "type": "string.file",
  },
  Object {
    "context": Object {
      "key": "database",
      "label": "database",
      "q": undefined,
      "v": "aJ@#LJ#@KL$KL@#sdf",
    },
    "message": "\\"database\\" does not exist",
    "path": Array [
      "database",
    ],
    "type": "string.file_exists",
  },
]
`;

exports[`Connections should perform basic validation 2`] = `
Array [
  Object {
    "context": Object {
      "key": "id",
      "label": "id",
      "value": 12,
    },
    "message": "\\"id\\" must be a string",
    "path": Array [
      "id",
    ],
    "type": "string.base",
  },
  Object {
    "context": Object {
      "key": "name",
      "label": "name",
    },
    "message": "\\"name\\" is required",
    "path": Array [
      "name",
    ],
    "type": "any.required",
  },
  Object {
    "context": Object {
      "key": "database",
      "label": "database",
      "q": undefined,
      "v": "/usr/foo",
    },
    "message": "\\"database\\" does not exist",
    "path": Array [
      "database",
    ],
    "type": "string.file_exists",
  },
]
`;

exports[`Connections should perform basic validation 3`] = `
Array [
  Object {
    "context": Object {
      "key": "name",
      "label": "name",
    },
    "message": "\\"name\\" is required",
    "path": Array [
      "name",
    ],
    "type": "any.required",
  },
  Object {
    "context": Object {
      "key": "database",
      "label": "database",
      "q": undefined,
      "v": "/usr/foo",
    },
    "message": "\\"database\\" does not exist",
    "path": Array [
      "database",
    ],
    "type": "string.file_exists",
  },
]
`;

exports[`Connections should perform basic validation 4`] = `
Array [
  Object {
    "context": Object {
      "key": "database",
      "label": "database",
      "q": undefined,
      "v": "/usr/foo",
    },
    "message": "\\"database\\" does not exist",
    "path": Array [
      "database",
    ],
    "type": "string.file_exists",
  },
]
`;

exports[`Connections should perform basic validation 5`] = `Array []`;

exports[`Connections should update a single connection 1`] = `
Array [
  Object {
    "id": "test-id-foo",
    "name": "test-connection-foo",
    "password": "test-password",
    "type": "sqlite",
  },
  Object {
    "id": "test-id-2",
    "name": "test-connection-2",
    "password": "test-password",
    "type": "sqlite",
  },
  Object {
    "id": "test-id-3",
    "name": "test-connection-3",
    "password": "test-password",
    "type": "sqlite",
  },
]
`;
